filterExprParser.go

```diff
@@ -0,0 +1,1160 @@
+// CompareOp                = "=" | "<>" | ">" | ">=" | "<" | "<="
+// CheckOp                  = "EXISTS" | ( "IS" [ "NOT" ] ( NULL | MISSING ) )
+// Field                    = OnePath { "." OnePath }
+// OnePath                  = ( PathFuncExpression | @String | @Ident ){ ArrayIndex }
+// ArrayIndex               = "[" [ @"-" ] @Int "]"
+// Value                    = @String
+// ConstFuncExpr            = ConstFuncNoArg | ConstFuncOneArg | ConstFuncTwoArgs
+// ConstFuncNoArg           = ConstFuncNoArgName "(" ")"
+// ConstFuncNoArgName       = "PI" | "E"
+// ConstFuncOneArg          = ConstFuncOneArgName "(" ConstFuncArgument ")"
+// ConstFuncOneArgName      = "ABS" | "ACOS"...
+// ConstFuncTwoArgs         = ConstFuncTwoArgsName "(" ConstFuncArgument "," ConstFuncArgument ")"
+// ConstFuncTwoArgsName     = "ATAN2" | "POW"
+// ConstFuncArgument        = Field | Value | ConstFuncExpr
+// ConstFuncArgumentRHS     = Value
+// PathFuncExpression       = OnePathFuncNoArg
+// OnePathFuncNoArg         = OnePathFuncNoArgName "(" ")"
+// OnePathFuncNoArgName     = "META"
+// BooleanFuncExpr          = BooleanFuncTwoArgs
+// BooleanFuncTwoArgs       = BooleanFuncTwoArgsName "(" ConstFuncArgument "," ConstFuncArgumentRHS ")"
+// BooleanFuncTwoArgsName   = "REGEXP_CONTAINS"
+
+type FilterExpression struct {
+	AndConditions []*FEAndCondition   `( @@ { "OR" @@ } )`
+	SubFilterExpr []*FilterExpression `{ "AND" @@ }`
+}
+
+func (fe *FilterExpression) String() string {
+	output := []string{}
+
+	first := true
+	for _, expr := range fe.AndConditions {
+		if first {
+			first = false
+		} else {
+			output = append(output, "OR")
+		}
+		output = append(output, expr.String())
+	}
+
+	for _, expr := range fe.SubFilterExpr {
+		if first {
+			first = false
+		} else {
+			output = append(output, "AND")
+		}
+		output = append(output, expr.String())
+	}
+
+	return strings.Join(output, " ")
+}
+
+// Outputs the head of the Expression match tree of which represents everything underneath
+func (f *FilterExpression) OutputExpression() (Expression, error) {
+	var outExpr OrExpr
+	for _, oneExpr := range f.AndConditions {
+		andExpr, err := oneExpr.OutputExpression()
+		if err != nil {
+			return outExpr, err
+		}
+		outExpr = append(outExpr, andExpr)
+	}
+
+	if len(f.SubFilterExpr) > 0 {
+		var combinedExpr AndExpr
+		combinedExpr = append(combinedExpr, outExpr)
+
+		for _, subFilterExpr := range f.SubFilterExpr {
+			subExpr, err := subFilterExpr.OutputExpression()
+			if err != nil {
+				return combinedExpr, err
+			}
+			combinedExpr = append(combinedExpr, subExpr)
+		}
+
+		return combinedExpr, nil
+	} else {
+		return outExpr, nil
+	}
+}
+
+type FEOpenParen struct {
+	Parens string `@"("`
+}
+
+func (feop *FEOpenParen) String() string {
+	return "("
+}
+
+type FECloseParen struct {
+	Parens string `@")"`
+}
+
+func (fecp *FECloseParen) String() string {
+	return ")"
+}
+
+type FEAndCondition struct {
+	OpenParens   []*FEOpenParen  `{ @@ }`
```

Comment:
is it necessary to support multiple, which would be redundant, parantheses here?

Reply:
Good catch. These are meant to used so that we can recursively detect if we have mismatched parenthesis. Will add code and test cases.